File: go/rz/engines.go
package rz

// PURE per-flow decision engines. Each returns a named-rule-backed decision.
// Nothing here touches the network, filesystem, caller, or an LLM.

type RuleDecision struct {
	RuleFired RuleId
	Decision  Decision
}

// ---------------------------------------------------------------------------
// Flow 3 + 5 base: failed subscription & mandate retry share retry semantics
// ---------------------------------------------------------------------------
func baseRetryDecision(input StoppingInput) RuleDecision {
	// (b) fraud -> suppress immediately
	if IsFraudFlagged(input.Reason) {
		return RuleDecision{RuleFired: RuleFraudSuppress, Decision: DecisionSuppress}
	}
	// (c) mandate_revoked -> escalate, never retry
	if IsMandateRevoked(input.Reason) {
		return RuleDecision{RuleFired: RuleMandateRevokedEsc, Decision: DecisionEscalate}
	}
	// (d) active PTP -> suppress until it passes
	if HasActivePromiseToPay(input.ActivePromiseToPayDate, input.Now) {
		return RuleDecision{RuleFired: RulePromiseToPaySuppress, Decision: DecisionSuppress}
	}
	// (e) exhausted retries -> escalate (no infinite loop)
	if AtMaxRetryAttempts(input.CurrentAttempt) || ExhaustedRetries(input.CurrentAttempt) {
		return RuleDecision{RuleFired: RuleExhaustAttemptsEscalate, Decision: DecisionEscalate}
	}
	// blast radius
	if AtTouchCap(input.TouchesForCustomer) {
		return RuleDecision{RuleFired: RuleMaxTouchesCap, Decision: DecisionAbandon}
	}
	// retryable?
	if !IsRetryable(input.Reason) {
		return RuleDecision{RuleFired: RuleNoOp, Decision: DecisionAbandon}
	}
	return RuleDecision{RuleFired: RuleTransientRetry, Decision: DecisionRetry}
}

func DecideFailedSubscription(input StoppingInput) RuleDecision {
	return baseRetryDecision(input)
}

// Flow 5: mandate retry — same base but respects NPCI retry window.
func DecideMandateRetry(input StoppingInput) RuleDecision {
	base := baseRetryDecision(input)
	if base.Decision == DecisionRetry {
		now := effectiveNow(input.Now)
		if !IsWithinMandateRetryWindow(now, input.RetryWindow) {
			return RuleDecision{RuleFired: RuleMandateRetryWindow, Decision: DecisionHold}
		}
		if !MandateRetryAttemptAllowed(input.CurrentAttempt, now, input.RetryWindow) {
			return RuleDecision{RuleFired: RuleMandateRetrySeq, Decision: DecisionHold}
		}
		return RuleDecision{RuleFired: RuleMandateRetrySeq, Decision: DecisionRetry}
	}
	return base
}

// ---------------------------------------------------------------------------
// Flow 2: checkout abandonment recovery
// ---------------------------------------------------------------------------
func DecideCheckoutAbandonment(input StoppingInput) RuleDecision {
	// disputed/error buckets that signal a real problem -> don't spam, escalate
	if input.Reason == ReasonCheckoutError {
		return RuleDecision{RuleFired: RuleCheckoutAbandon, Decision: DecisionEscalate}
	}
	// Don't chase first-time visitors without evidence of intent (spam guard)
	if !IsRepeatVisitor(input.Visits) {
		return RuleDecision{RuleFired: RuleRepeatVisitorOnly, Decision: DecisionAbandon}
	}
	// blast radius
	if AtCheckoutReminderCap(input.TouchesForCustomer) || AtTouchCap(input.TouchesForCustomer) {
		return RuleDecision{RuleFired: RuleMaxTouchesCap, Decision: DecisionAbandon}
	}
	// If eligible, offer a cart incentive on FIRST touch to recover.
	if input.TouchesForCustomer == 0 && CartEligibleForIncentive(input.CartValue) {
		return RuleDecision{RuleFired: RuleCartIncentive, Decision: DecisionContact}
	}
	return RuleDecision{RuleFired: RuleCheckoutReminder, Decision: DecisionContact}
}

// ---------------------------------------------------------------------------
// Flow 4: B2B receivables chaser
// ---------------------------------------------------------------------------
func DecideB2BReceivables(input StoppingInput) RuleDecision {
	// Disputed invoice -> hold automated chasing, route to collections human
	if IsMandateRevoked(input.Reason) {
		return RuleDecision{RuleFired: RuleDisputeHold, Decision: DecisionHold}
	}
	if input.Reason == ReasonDisputedReceivable {
		return RuleDecision{RuleFired: RuleDisputeHold, Decision: DecisionHold}
	}

	days := input.OverdueDays
	if days == 0 {
		days = 0
	}
	tier := ReceivableTier(days)
	action := ReceivableAction(tier)

	// blast radius: don't over-chase a debtor
	if AtTouchCap(input.TouchesForCustomer) {
		return RuleDecision{RuleFired: RuleMaxTouchesCap, Decision: DecisionAbandon}
	}
	if action == "none" {
		return RuleDecision{RuleFired: RuleNoOp, Decision: DecisionNone}
	}
	if action == "remind" || action == "smtp" {
		return RuleDecision{RuleFired: RuleInvoiceReminder, Decision: DecisionContact}
	}
	// dunning / legal -> escalate to collections
	return RuleDecision{RuleFired: RuleInvoiceEscalateDun, Decision: DecisionEscalate}
}

// ---------------------------------------------------------------------------
// Flow 1: payment degradation -> root cause -> intervention
// ---------------------------------------------------------------------------
func DecidePaymentDegradation(input StoppingInput) RuleDecision {
	if input.Reason == ReasonFraudFlagged {
		return RuleDecision{RuleFired: RuleFraudSuppress, Decision: DecisionSuppress}
	}
	if input.Reason == ReasonIssuerDown {
		return RuleDecision{RuleFired: RuleDegradationEscalate, Decision: DecisionEscalate}
	}
	rate := input.RollingSuccessRate
	if rate == 0 {
		rate = 1
	}
	if SuccessRateBelowThreshold(rate) {
		return RuleDecision{RuleFired: RuleDegradationAlert, Decision: DecisionContact}
	}
	return RuleDecision{RuleFired: RuleDegradationRecover, Decision: DecisionRecover}
}

// ---------------------------------------------------------------------------
// Flow 6: Hinglish voice recovery
// ---------------------------------------------------------------------------
func DecideHinglishVoice(input StoppingInput) RuleDecision {
	if IsMandateRevoked(input.Reason) {
		return RuleDecision{RuleFired: RuleMandateRevokedEsc, Decision: DecisionEscalate}
	}
	if IsFraudFlagged(input.Reason) {
		return RuleDecision{RuleFired: RuleFraudSuppress, Decision: DecisionSuppress}
	}
	// Regulatory/TRAI quiet hours + DNC
	if IsDoNotCall(input.DNCFlag, effectiveNow(input.Now)) {
		return RuleDecision{RuleFired: RuleVoiceDoNotCall, Decision: DecisionAbandon}
	}
	if input.Reason == ReasonVoiceUnreachable {
		return RuleDecision{RuleFired: RuleVoiceRetry, Decision: DecisionContact}
	}
	if AtVoiceCallCap(input.TouchesForCustomer) {
		return RuleDecision{RuleFired: RuleVoiceEscalateHuman, Decision: DecisionEscalate}
	}
	if AtTouchCap(input.TouchesForCustomer) {
		return RuleDecision{RuleFired: RuleMaxTouchesCap, Decision: DecisionAbandon}
	}
	return RuleDecision{RuleFired: RuleVoiceCall, Decision: DecisionContact}
}

// ---------------------------------------------------------------------------
// Flow 7: promise-to-pay tracker
// ---------------------------------------------------------------------------
func DecidePromiseToPay(input StoppingInput) RuleDecision {
	t := effectiveNow(input.Now)

	if input.Reason == ReasonPtpBroken || input.Reason == ReasonPtpMissed {
		// Customer broke/missed a PTP -> escalate (human/collections), don't spam
		return RuleDecision{RuleFired: RulePtpMissedEscalate, Decision: DecisionEscalate}
	}
	if input.Reason == ReasonFraudFlagged {
		return RuleDecision{RuleFired: RuleFraudSuppress, Decision: DecisionSuppress}
	}

	if input.ActivePromiseToPayDate == nil {
		// No PTP on file -> if this is a ptp_committed event lacking a date, hold
		return RuleDecision{RuleFired: RulePtpSchedule, Decision: DecisionHold}
	}

	ptpDate := *input.ActivePromiseToPayDate

	// Compliance: large PTPs need supervisor sign-off before automated follow-up
	if PtpNeedsSupervisor(input.Amount) {
		return RuleDecision{RuleFired: RulePtpMissedEscalate, Decision: DecisionEscalate}
	}

	if PtpDateNotYet(t, ptpDate) {
		if PtpReminderDue(t, ptpDate) {
			return RuleDecision{RuleFired: RulePtpReminderBefore, Decision: DecisionContact}
		}
		// Suppress all other touches until the PTP date passes (rule d)
		return RuleDecision{RuleFired: RulePtpSuppress, Decision: DecisionSuppress}
	}

	if PtpMissed(t, ptpDate) {
		return RuleDecision{RuleFired: RulePtpFollowup, Decision: DecisionContact}
	}

	return RuleDecision{RuleFired: RuleNoOp, Decision: DecisionNone}
}

// DecideForFlow dispatches by flow.
func DecideForFlow(flow FlowType, input StoppingInput) RuleDecision {
	switch flow {
	case FlowFailedSubscription:
		return DecideFailedSubscription(input)
	case FlowMandateRetry:
		return DecideMandateRetry(input)
	case FlowCheckoutAbandonment:
		return DecideCheckoutAbandonment(input)
	case FlowB2BReceivables:
		return DecideB2BReceivables(input)
	case FlowPaymentDegradation:
		return DecidePaymentDegradation(input)
	case FlowHinglishVoice:
		return DecideHinglishVoice(input)
	case FlowPromiseToPay:
		return DecidePromiseToPay(input)
	}
	return RuleDecision{RuleFired: RuleNoOp, Decision: DecisionNone}
}

File: src/decisions/engines.ts
// PURE per-flow decision engines. Each returns a named-rule-backed decision.
// Nothing here touches the network, filesystem, caller, or an LLM.

import { FlowType, ReasonBucket } from '../flows/types';
import {
  StoppingInput,
  atMaxRetryAttempts,
  atTouchCap,
  exhaustedRetries,
  hasActivePromiseToPay,
  isFraudFlagged,
  isMandateRevoked,
  isRetryable,
  isWithinMandateRetryWindow,
  mandateRetryAttemptAllowed,
  atCheckoutReminderCap,
  isRepeatVisitor,
  cartEligibleForIncentive,
  receivableTier,
  receivableAction,
  atVoiceCallCap,
  isDoNotCall,
  ptpReminderDue,
  ptpDateNotYet,
  ptpMissed,
  ptpNeedsSupervisor,
  successRateBelowThreshold,
} from './rules';
import { EvaluatedStep } from '../flows/types';

export type RuleDecision = { ruleFired: string; decision: 'retry' | 'contact' | 'suppress' | 'escalate' | 'abandon' | 'hold' | 'recover' | 'none' };

// ---------------------------------------------------------------------------
// Flow 3 + 5 base: failed subscription & mandate retry share retry semantics
// ---------------------------------------------------------------------------
function baseRetryDecision(input: StoppingInput): RuleDecision {
  const { reason, currentAttempt, touchesForCustomer, activePromiseToPayDate, now } = input;

  // (b) fraud -> suppress immediately
  if (isFraudFlagged(reason)) return { ruleFired: 'fraud_suppress', decision: 'suppress' };
  // (c) mandate_revoked -> escalate, never retry
  if (isMandateRevoked(reason)) return { ruleFired: 'mandate_revoked_escalate', decision: 'escalate' };
  // (d) active PTP -> suppress until it passes
  if (hasActivePromiseToPay({ activePromiseToPayDate, now })) {
    return { ruleFired: 'promise_to_pay_suppress', decision: 'suppress' };
  }
  // (e) exhausted retries -> escalate (no infinite loop)
  if (atMaxRetryAttempts(currentAttempt) || exhaustedRetries(currentAttempt)) {
    return { ruleFired: 'exhaust_attempts_escalate', decision: 'escalate' };
  }
  // blast radius
  if (atTouchCap(touchesForCustomer)) return { ruleFired: 'max_touches_cap', decision: 'abandon' };
  // retryable?
  if (!isRetryable(reason)) return { ruleFired: 'no_op', decision: 'abandon' };
  return { ruleFired: 'transient_retry', decision: 'retry' };
}

export function decideFailedSubscription(input: StoppingInput): RuleDecision {
  return baseRetryDecision(input);
}

// Flow 5: mandate retry — same base but respects NPCI retry window.
export function decideMandateRetry(input: StoppingInput): RuleDecision {
  const base = baseRetryDecision(input);
  if (base.decision === 'retry') {
    const now = input.now ?? Date.now();
    if (!isWithinMandateRetryWindow(now, input.retryWindow)) {
      return { ruleFired: 'mandate_retry_window', decision: 'hold' };
    }
    if (!mandateRetryAttemptAllowed(input.currentAttempt, now, input.retryWindow)) {
      return { ruleFired: 'mandate_retry_seq', decision: 'hold' };
    }
    return { ruleFired: 'mandate_retry_seq', decision: 'retry' };
  }
  return base;
}

// ---------------------------------------------------------------------------
// Flow 2: checkout abandonment recovery
// ---------------------------------------------------------------------------
export function decideCheckoutAbandonment(input: StoppingInput): RuleDecision {
  const { touchesForCustomer, cartValue, visits, reason } = input;

  // disputed/error buckets that signal a real problem -> don't spam, escalate
  if (reason === 'checkout_error') {
    return { ruleFired: 'checkout_abandon', decision: 'escalate' };
  }
  // Don't chase first-time visitors without evidence of intent (spam guard)
  if (!isRepeatVisitor(visits)) {
    return { ruleFired: 'repeat_visitor_only', decision: 'abandon' };
  }
  // blast radius
  if (atCheckoutReminderCap(touchesForCustomer) || atTouchCap(touchesForCustomer)) {
    return { ruleFired: 'max_touches_cap', decision: 'abandon' };
  }
  // If eligible, offer a cart incentive on FIRST touch to recover.
  if (touchesForCustomer === 0 && cartEligibleForIncentive(cartValue ?? 0)) {
    return { ruleFired: 'cart_incentive', decision: 'contact' };
  }
  return { ruleFired: 'checkout_reminder', decision: 'contact' };
}

// ---------------------------------------------------------------------------
// Flow 4: B2B receivables chaser
// ---------------------------------------------------------------------------
export function decideB2BReceivables(input: StoppingInput): RuleDecision {
  const { reason, overdueDays, touchesForCustomer } = input;

  // Disputed invoice -> hold automated chasing, route to collections human
  if (isMandateRevoked(reason)) return { ruleFired: 'dispute_hold', decision: 'hold' };
  if (reason === 'disputed_receivable') return { ruleFired: 'dispute_hold', decision: 'hold' };

  const days = input.overdueDays ?? 0;
  const tier = receivableTier(days);
  const action = receivableAction(tier);

  // blast radius: don't over-chase a debtor
  if (atTouchCap(touchesForCustomer)) {
    return { ruleFired: 'max_touches_cap', decision: 'abandon' };
  }
  if (action === 'none') return { ruleFired: 'no_op', decision: 'none' };
  if (action === 'remind') return { ruleFired: 'invoice_reminder', decision: 'contact' };
  if (action === 'smtp') return { ruleFired: 'invoice_reminder', decision: 'contact' };
  // dunning / legal -> escalate to collections
  return { ruleFired: 'invoice_escalate_dunning', decision: 'escalate' };
}

// ---------------------------------------------------------------------------
// Flow 1: payment degradation -> root cause -> intervention
// ---------------------------------------------------------------------------
export function decidePaymentDegradation(input: StoppingInput): RuleDecision {
  const { reason } = input;

  if (reason === 'fraud_flagged') return { ruleFired: 'fraud_suppress', decision: 'suppress' };
  if (reason === 'issuer_down') {
    return { ruleFired: 'degradation_escalate', decision: 'escalate' };
  }
  if (successRateBelowThreshold(input.rollingSuccessRate ?? 1)) {
    return { ruleFired: 'degradation_alert', decision: 'contact' };
  }
  return { ruleFired: 'degradation_recover', decision: 'recover' };
}

// ---------------------------------------------------------------------------
// Flow 6: Hinglish voice recovery
// ---------------------------------------------------------------------------
export function decideHinglishVoice(input: StoppingInput): RuleDecision {
  const { reason, touchesForCustomer, now, dncFlag } = input;

  if (isMandateRevoked(reason)) return { ruleFired: 'mandate_revoked_escalate', decision: 'escalate' };
  if (isFraudFlagged(reason)) return { ruleFired: 'fraud_suppress', decision: 'suppress' };
  // Regulatory/TRAI quiet hours + DNC
  if (isDoNotCall({ dncFlag, now })) return { ruleFired: 'voice_do_not_call', decision: 'abandon' };
  if (reason === 'voice_unreachable') return { ruleFired: 'voice_retry', decision: 'contact' };
  if (atVoiceCallCap(touchesForCustomer)) {
    return { ruleFired: 'voice_escalate_human', decision: 'escalate' };
  }
  if (atTouchCap(touchesForCustomer)) return { ruleFired: 'max_touches_cap', decision: 'abandon' };
  return { ruleFired: 'voice_call', decision: 'contact' };
}

// ---------------------------------------------------------------------------
// Flow 7: promise-to-pay tracker
// ---------------------------------------------------------------------------
export function decidePromiseToPay(input: StoppingInput): RuleDecision {
  const { reason, activePromiseToPayDate, now, amount } = input;
  const t = now ?? Date.now();

  if (reason === 'ptp_broken') {
    // Customer broke a PTP -> escalate (human/collections), don't spam
    return { ruleFired: 'ptp_missed_escalate', decision: 'escalate' };
  }
  if (reason === 'ptp_missed') {
    return { ruleFired: 'ptp_missed_escalate', decision: 'escalate' };
  }
  if (reason === 'fraud_flagged') return { ruleFired: 'fraud_suppress', decision: 'suppress' };

  if (activePromiseToPayDate == null) {
    // No PTP on file -> if this is a ptp_committed event lacking a date, hold
    return { ruleFired: 'ptp_schedule', decision: 'hold' };
  }

  // Compliance: large PTPs need supervisor sign-off before automated follow-up
  if (ptpNeedsSupervisor(amount ?? 0)) {
    return { ruleFired: 'ptp_missed_escalate', decision: 'escalate' };
  }

  if (ptpDateNotYet(t, activePromiseToPayDate)) {
    if (ptpReminderDue(t, activePromiseToPayDate)) {
      return { ruleFired: 'ptp_reminder_before', decision: 'contact' };
    }
    // Suppress all other touches until the PTP date passes (rule d)
    return { ruleFired: 'ptp_suppress', decision: 'suppress' };
  }

  if (ptpMissed(t, activePromiseToPayDate)) {
    return { ruleFired: 'ptp_followup', decision: 'contact' };
  }

  return { ruleFired: 'no_op', decision: 'none' };
}

// Dispatch by flow
export function decideForFlow(flow: FlowType, input: StoppingInput): RuleDecision {
  switch (flow) {
    case 'failed_subscription': return decideFailedSubscription(input);
    case 'mandate_retry': return decideMandateRetry(input);
    case 'checkout_abandonment': return decideCheckoutAbandonment(input);
    case 'b2b_receivables': return decideB2BReceivables(input);
    case 'payment_degradation': return decidePaymentDegradation(input);
    case 'hinglish_voice': return decideHinglishVoice(input);
    case 'promise_to_pay': return decidePromiseToPay(input);
  }
}

export type { EvaluatedStep };
